fs/dafsserver/dafsserver.cpp

```diff
@@ -2915,93 +2915,95 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
+                        msg.read(left);
+                        msg.clear();
+                        try
                         {
-                            size32_t rd = avail>sizeof(fbuf)?sizeof(fbuf):avail;
-                            try
-                            {
-                                socket->read(fbuf, rd); // don't need timeout here
-                                avail -= rd;
-                            }
-                            catch (IException *e)
+                            msgWritePtr = (byte *)msg.reserveTruncate(left);
+                        }
+                        catch (IException *e)
+                        {
+                            EXCLOG(e,"notifySelected(1)");
+                            e->Release();
+                            left = 0;
+                            // if too big then suggest corrupted packet, try to consume
+                            // JCSMORE this seems a bit pointless, and it used to only read last 'avail',
+                            // which is not necessarily everything that was sent
+                            char fbuf[1024];
+                            while (true)
                             {
-                                EXCLOG(e,"notifySelected(2) flush");
-                                e->Release();
-                                break;
+                                try
+                                {
+                                    size32_t szRead;
+                                    sock->read(fbuf, 1, 1024, szRead);
+                                }
+                                catch (IException *e)
+                                {
+                                    EXCLOG(e,"notifySelected(2)");
+                                    e->Release();
+                                    break;
+                                }
                             }
                         }
-                        avail = 0;
-                        left = 0;
+                        if (0 == left)
+                        {
+                            gotSize = false;
+                            msg.clear();
+                            parent->onCloseSocket(this, 5);
+                            return true;
+                        }
                     }
                 }
-            }
-            size32_t toread = left>avail?avail:left;
-            if (toread)
-            {
-                try
-                {
-                    socket->read(msg.reserve(toread), toread);  // don't need timeout here
-                }
-                catch (IException *e)
+                else // left represents length of message remaining to receive
                 {
-                    EXCLOG(e,"notifySelected(3)");
-                    e->Release();
-                    toread = left;
-                    msg.clear();
+                    size32_t szRead;
+                    sock->read(msgWritePtr, 1, left, szRead);
```

Comment:
can we make min_size here left ?  Either we get as much as we know we need or we time out ?

Reply:
We don't want to block the select handler, so that other sockets are serviced.
This reads as much as it can then returns, to be notified to read more/rest when notified again.